crates/blockchain/blockchain.rs

```diff
@@ -159,6 +142,163 @@ pub fn validate_block(
     Ok(())
 }
 
+pub fn is_canonical(
+    store: &Store,
+    block_number: BlockNumber,
+    block_hash: BlockHash,
+) -> Result<bool, StoreError> {
+    match store.get_canonical_block_hash(block_number)? {
+        Some(hash) if hash == block_hash => Ok(true),
+        _ => Ok(false),
+    }
+}
+
+/// Applies new fork choice data to the current blockchain. It performs validity checks:
+/// - The finalized, safe and head hashes must correspond to already saved blocks.
+/// - The saved blocks should be in the correct order (finalized <= safe <= head).
+/// - They must be connected.
+///
+/// After the validity checks, the canonical chain is updated so that all head's ancestors
+/// and itself are made canonical.
+///
+/// If the fork choice state is applied correctly, the head block header is returned.
+pub fn apply_fork_choice(
```

Comment:
nit: should we move this and related logic to a new module, maybe `fork_choice`?